velox/exec/GroupingSet.cpp

```diff
@@ -850,6 +855,31 @@ void GroupingSet::ensureInputFits(const RowVectorPtr& input) {
           0, outOfLineBytes - (rowsToSpill * outOfLineBytesPerRow)));
 }
 
+void GroupingSet::ensureOutputFits() {
+  // If spilling has already been triggered on this operator, then we don't need
+  // to reserve memory for the output as we can't reclaim much memory from this
+  // operator itself. The output processing can reclaim memory from the other
+  // operator or query through memory arbitration.
+  if (isPartial_ || spillConfig_ == nullptr || hasSpilled()) {
+    return;
+  }
+
+  // Test-only spill path.
+  if (spillConfig_->testSpillPct > 0 &&
+      (folly::hasher<uint64_t>()(++spillTestCounter_)) % 100 <=
+          spillConfig_->testSpillPct) {
+    spill(RowContainerIterator{});
+    return;
+  }
+
+  const uint64_t outputBufferSizeToReserve =
+      queryConfig_->preferredOutputBatchBytes() * 1.2;
+  if (pool_.maybeReserve(outputBufferSizeToReserve)) {
+    return;
+  }
+  spill(RowContainerIterator{});
```

Comment:
Do we want to check if after spilling there is enough memory space for output?

Reply:
Not much we could do here. If we run out of memory on output processing, either arbitrator reclaims memory from the other query or operator or this query just fails. Not much we can do here.